doc/design.md

```diff
@@ -0,0 +1,226 @@
+
+Needs: dsn
+
+# Solution Strategy
+
+Row-level security is not part of the Exasol core database. However with the [Virtual Schema](https://github.com/exasol/virtual-schemas) interface lends itself nicely to what we need to implement RLS as a plug-in.
+
+At its core Exasol Virtual Schema's are a query rewriter, not unlike views. You put in a query and it passes a modified query back that is then immediately executed by the database core instead of the original one. This is exactly what we need to implement RLS.
+
+## Requirement Overview
+
+Please refer to the [System Requirement Specification](system_requirements.md) for user-level requirements.
+
+# Building Blocks
+
+## Row Level Security Adapter
+`dsn~row-level-security-adapter~1`
+
+The `RowLevelSecurityAdapter` is a Virtual Schema Adapter that modifies queries it receives so that they only return data the current user is allowed to see.
+
+Comment:
+
+Since this component implements the Virtual Schema Adapter Interface, it is also the entry point for the RLS plug-in.
+
+Covers:
+
+* `req~rows-users-are-allowed-to-read~1`
+
+Needs: impl
+
+## `QueryRewriter`
+`dsn~query-rewriter~1`
+
+The `QueryRewriter` is responsible for modifying the original query in order to inject row filters.
+
+Covers:
+
+* `req~rows-users-are-allowed-to-read~1`
+
+Needs: impl
+
+## `UserInformation`
+
+The `UserInformation` keeps details about the users role assignments.
+
+Covers:
+
+* `req~user-roles~1`
+
+# Runtime
+
+## `RowLevelSecurityAdapter` Reads Custom Properties
+`dsn~rls-adapter-reads-custom-properties~1`
+
+The `RowLevelSecurityAdapter` reads the following custom properties:
+
+1. `ROLE_ASSIGNMENT_TABLE`: name of the table that contains the mapping of users to the roles the have
+
+Covers:
+
+* `req~user-roles~1`
+
+Needs: impl, utest, itest
+
+## `UserInformation` Read User Roles
+`dsn~user-information-reads-user-roles`
+
+The `UserInformation` reads the current user's roles from a table with the following layout:
+
+* `exa_user_name VARCHAR(128)` (see [https://docs.exasol.com/sql_references/basiclanguageelements.htm#SQL_Identifier]("SQL Identifier") in the Exasol documentation)
+* `exa_role_mask DECIMAL(20,0)`
+
+Covers:
+
+* `req~user-roles~1`
+
+Needs: impl, utest, itest
+
+## `UserInformation` Caches User Roles
+`dsn~user-information-caches-user-roles`
+
+The `UserInformation` caches the roles a user has.
+
+Rationale:
+
+This speeds up subsequent row access checks because the users roles need to be determined only once.
+
+Covers:
+
+* `req~user-roles~1`
+
+Needs: impl, utest
+
+## Query Rewriter Determines User Roles
+
+## Query Rewriter Identifies Protected Tables
+`dsn~query-rewriter-identifies-protected-tables~1`
+
+The Query Rewriter identifies a table as protected with row-level security, if that table has a column named `exa_row_roles`.
+
+Covers:
+
+* `req~rows-users-are-allowed-to-read~1`
+
+Needs: impl, utest, itest
+
+## Query Rewriter Replaces Tables
```

Comment:
That sounds a little bit misleading. As if we delete the origin table and replace it with something else